jobs/ml_jobs/retrieval_vector/app/retrieval/core/reranker.py

```diff
@@ -0,0 +1,143 @@
+from typing import Optional
+
+import numpy as np
+import pyarrow as pa
+from docarray import Document, DocumentArray
+from lancedb.rerankers import Reranker
+
+from app.logger import logger
+
+
+class UserReranker(Reranker):
+    """
+    A user-based reranker class that reranks results based on the user's vector.
+
+    Attributes:
+        weight (float): The weight to balance between distance and user embedding similarity.
+        user_docs (DocumentArray): Document array containing user vectors.
+    """
+
+    def __init__(
+        self,
+        weight: float,
+        user_docs: DocumentArray,
+        return_score: str = "all",
+        vector_column_name: str = "vector",
+    ) -> None:
+        """
+        Initializes the UserReranker with a given weight and user document vectors.
+
+        Args:
+            weight (float): Weight to balance between vector distance and user vector similarity.
+            user_docs (DocumentArray): Document array containing user embedding vectors.
+            return_score (str): Type of score to return (e.g., "all").
+            vector_column_name (str): Column field name.
+        """
+        super().__init__(return_score)
+        self.weight = weight
+        self.user_docs = user_docs
+        self.vector_column_name = vector_column_name
+
+    def user_vector(self, user_id: str) -> Optional[Document]:
+        """
+        Retrieves the user vector from the document array based on the given user ID.
+
+        Args:
+            user_id (str): The ID of the user.
+
+        Returns:
+            Optional[Document]: The user's document embedding, or None if not found.
+        """
+        try:
+            return self.user_docs[user_id]
+        except KeyError:
+            return None
+
+    def _user_recommendation(self, vector_results: pa.Table, user_id: str) -> pa.Table:
+        """
+        Re-ranks vector results based on user-specific vector embeddings.
+
+        Args:
+            vector_results (pa.Table): Vector search results as a PyArrow table.
+            user_id (str): The user ID whose embedding should be used for reranking.
+
+        Returns:
+            pa.Table: The reranked results based on user vector similarity and original vector distance.
+        """
+        user_doc = self.user_vector(user_id)
+
+        if user_doc:
+            scores = np.dot(
+                vector_results[self.vector_column_name].to_pylist(), -user_doc.embedding
+            )
```

Comment:
pourquoi ne pas utiliser to_numpy() directement au lieu de to_pylist()?

Reply:
to_numpy() not working here